File: algorithms/Go/searching/binary-search.go
package searching

import (
	"fmt"
	"math/rand"
	"time"
)

func binarySearch(arr []int, elem int) (int, bool) {
	var (
		start = 0
		end   = len(arr) - 1
	)

	for start <= end {
		mid := start + (end-start)/2
		if elem == arr[mid] {

			return mid, true
		} else if elem < arr[mid] {
			end = mid - 1
		} else {
			start = mid + 1
		}
	}
	return -1, false

}

func RunBinarySearch() {
	rand.Seed(time.Now().UnixNano())
	arr := []int{1, 2, 4, 7, 8, 9, 12, 14, 17, 18, 20, 22, 23, 24, 26, 27, 29}
	elem := rand.Intn(20)

	fmt.Printf("Array: %v\nNumber to be searched: %v\n", arr, elem)
	index, found := binarySearch(arr, elem)
	if !found {
		fmt.Printf("Number not found!\n")
	} else {
		fmt.Printf("Number found at index %v", index)
	}

}

File: algorithms/Java/searching/binary-search.java
// Algorithm BinarySearch Iterative method
/*binarySearch(arr, x, low, high)
        repeat till low = high
               mid = (low + high)/2
                   if (x == arr[mid])
                   return mid
   
                   else if (x > arr[mid]) // x is on the right side
                       low = mid + 1
   
                   else                  // x is on the left side
                       high = mid - 1
 */
// Time Complexity : O(log(n)) 

public class BinarySearch {

    static int binarySearch(int arr[], int key) 
    {
        int start = 0;
        int end = arr.length - 1;
        
        while (start <= end) {
            // We use (start + (end - start)/2) rather than using (start + end)/2 to avoid
            // arithmetic overflow.
            // Arithmetic overflow is the situation when the value of a variable increases
            // beyond the maximum value of the memory location, and wraps around.
            int mid = start + (end - start) / 2; // optimised way

            if (arr[mid] == key)// key element is found at the middle of the array
                return mid;

            else if (arr[mid] < key) {// so the key lies in the right hand side of array
                start = mid + 1;
            }

            else {// so the key lies in the left subarray
                end = mid - 1;
            }
        }
        // we reach here when the key element is not present
        return -1;
    }

    public static void main(String[] args) 
    {

        int arr[] = { 1, 3, 4, 5, 6 };

        /*
         * List<ArrayList<Integer>> arr = new ArrayList<>();
         * arr.add(new ArrayList<Integer>(Arrays.asList( 1, 3, 4, 5, 6 )));
         */
        int key = 4; // element to search
        int index = binarySearch(arr, key);
        if (index == -1) {
            System.out.println("key element not found");
        }
        else {
            System.out.println("key element found at index :" + index);
        }

    }
}
